### PR TITLE
New data source `azurerm_images`

### LINKED ISSUE #8417: Support for filtering azurerm_image data resources by tags
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure Machine Images can't be renamed, the only option available to mutate an image metadata as it progresses through a validation pipeline are the tags, we use this for marking the state of an image (i.e. RELEASE_STATUS=RC or RELEASE_STATUS=GA). The current regex-on-name filter means that once an image is published, it's available for use, which isn't desirable as it may not have passed testing yet.

### New or Affected Resource(s)

* Data Source: azurerm_image

### Potential Terraform Configuration

```hcl
data "azurerm_image" "search" {
  name                = "search-api"
  resource_group_name = "packerimages"
  tags = {
    RELEASE_STATUS = "GA"
  }
}
```

### References

* this would be very similar to the AWS AMI data source: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/ami

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Partially helps #8417

### CODE REVIEW

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @ArcturusZhang, thanks for this PR!

Overall this looks great and I have just a few minor suggestions, however I'm not convinced that it addresses the linked issue? Could you look at adding the `tags_filter` attribute to the `azurerm_image` resource too?

**Comment by ArcturusZhang** [comment]
Hi @manicminer thanks for the review. I will get the comments resolved soon.

And to clarify, 
> Could you look at adding the tags_filter attribute to the azurerm_image resource too?
Do you mean that we need to add `tags_filter` to the `azurerm_image` data source?

**Comment by manicminer** [comment]
> Do you mean that we need to add `tags_filter` to the `azurerm_image` data source?

Sorry, that is indeed what I meant :)

File: azurerm/internal/services/compute/images_data_source.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		return fmt.Errorf("retrieving Images (Resource Group %q): %+v", resourceGroup, err)
+	}
+
+	images, err := flattenImages(ctx, resp, filterTags)
+	if err != nil {
+		return fmt.Errorf("retrieving Images (Resource Group %q): %+v", resourceGroup, err)
+	}
+	if len(images) == 0 {
+		return fmt.Errorf("unable to find any images")
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/images_data_source.go:139]: Done

File: azurerm/internal/services/compute/images_data_source.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	filterTags := tags.Expand(d.Get("tags_filter").(map[string]interface{}))
+
+	resp, err := client.ListByResourceGroupComplete(ctx, resourceGroup)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response().Response) {
+			return fmt.Errorf("Images (Resource Group %q) was not found", resourceGroup)
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/images_data_source.go:129]: Done

File: azurerm/internal/services/compute/images_data_source.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		return fmt.Errorf("retrieving Images (Resource Group %q): %+v", resourceGroup, err)
+	}
+
+	images, err := flattenImages(ctx, resp, filterTags)
+	if err != nil {
+		return fmt.Errorf("retrieving Images (Resource Group %q): %+v", resourceGroup, err)
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/images_data_source.go:136]: Done

File: azurerm/internal/services/compute/images_data_source.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	d.Set("resource_group_name", resourceGroup)
+
+	if err := d.Set("images", images); err != nil {
+		return fmt.Errorf("setting `images`: %+v", err)
+	}
+
+	return nil
+}
+
+func flattenImages(ctx context.Context, iterator compute.ImageListResultIterator, filterTags map[string]*string) ([]interface{}, error) {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/images_data_source.go:153]: Done

File: azurerm/internal/services/compute/images_data_source.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+	results := make([]interface{}, 0)
+
+	var err error
+	for ; iterator.NotDone(); err = iterator.NextWithContext(ctx) {
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/images_data_source.go:157]: Done

**Comment by ArcturusZhang** [comment]
Hi @manicminer I have made some changes, please have a look.

I do not think we should do too many things in one PR, therefore I might create another PR to add `tags_filter` for the azurerm_image data source. And I changed the description so that this PR will not close the coresponding issue.

File: azurerm/internal/services/compute/tests/images_data_source_test.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+			},
+			{
+				Config: testAccAzureRMImage_standaloneImage_provision(data, userName, password, hostName, storageType, ""),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMImageExists("azurerm_image.test", true),
+				),
+			},
+			{
+				Config:      testAccDataSourceImages_tagsFilterError(data, userName, password, hostName, storageType),
+				ExpectError: regexp.MustCompile("unable to find any images"),
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply azurerm/internal/services/compute/tests/images_data_source_test.go:83]: Oh sorry I missed this one.

**Comment by ArcturusZhang** [comment]
Hi @manicminer I have resolved the comment, could you please have a look again? thanks

**Review by manicminer — APPROVED** [review]

**Comment by manicminer** [comment]
Test results:

Screenshot 2020-11-02 at 12 46 36

**Pull request merged**